File: covcon/benchmarks/instrumented/Day/src/test/DeadlockMonitor.java
package test;


import java.lang.management.ManagementFactory;
import java.lang.management.ThreadInfo;
import java.lang.management.ThreadMXBean;

public class DeadlockMonitor extends Thread {
	
	private ThreadMXBean threadMgr = ManagementFactory.getThreadMXBean();
	
	/**
	 * Checks periodically for deadlocks and reports them.
	 */
	public void run() {
		while (true) {
			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
			long[] threads = threadMgr.findDeadlockedThreads();
			if (threads != null) {  // found a deadlock
				System.out.println("\n ============================= Deadlock =============================");
				ThreadInfo[] threadInfos = threadMgr.getThreadInfo(threads, true, true);
				for (int i = 0; i < threadInfos.length; i++) {
					ThreadInfo threadInfo = threadInfos[i];
					System.out.println(threadInfo);
				}
				System.out.println("=====================================================================");
				System.exit(0);
			}
		}
	}

}

File: covcon/src/contege/seqexec/DeadlockMonitor.scala
package contege.seqexec

import java.lang.management._
import contege.ClassTester
import contege.Config
import scala.collection.mutable.Map
import contege.Finalizer
import contege.GlobalState

class DeadlockMonitor(config: Config, global: GlobalState) extends Thread {
	
	private val threadMgr = ManagementFactory.getThreadMXBean;
	
	/**
	 * Checks periodically for deadlocks and reports them.
	 */
	override def run() {
		while (true) {
			Thread.sleep(1000);
			val threads = threadMgr.findDeadlockedThreads
			if (threads != null) {  // found a deadlock
			    config.checkerListeners.foreach(_.appendResultMsg("\n =========Found a thread safety violation (deadlock)! ========" ))
			    config.checkerListeners.foreach(_.appendResultMsg("Exception Found : Deadlock"))
				val threadInfos = threadMgr.getThreadInfo(threads, true, true)
				threadInfos.foreach(info => config.checkerListeners.foreach(_.appendResultMsg(info.toString)))
				config.checkerListeners.foreach(_.appendResultMsg("=====================================================================" ))
				
				global.finalizer.finalizeAndExit(true)
			}
		}
	}

}